packages/ag-charts-community/src/chart/legend/legendManager.ts

```diff
@@ -0,0 +1,151 @@
+    }
+
+    public restoreMemento(_version: string, _mementoVersion: string, memento: LegendDataMemento | undefined) {
+        const { legendDataMap } = this;
+        memento?.forEach((datum) => {
+            const { seriesId, data } = this.getRestoredDatum(datum) ?? {};
+
+            if (!(seriesId && data)) {
+                return;
+            }
+
+            legendDataMap.set(seriesId, data);
+        });
+
+        this.update();
+    }
+
+    private getRestoredDatum(datum: AgInitialStateLegendOptions) {
+        const { seriesId, itemId, visible } = datum;
+
+        if (!datum.seriesId && !datum.itemId) {
+            return;
+        }
+
+        if (seriesId) {
+            const legendData = this.legendDataMap.get(seriesId) ?? [];
+
+            const data = legendData.map((d) =>
+                d.seriesId === seriesId && (!datum.itemId || d.itemId === datum.itemId) ? { ...d, enabled: visible } : d
+            );
+
+            return { seriesId, data };
+        }
+
+        if (itemId) {
+            const legendData = this.getData();
+
+            for (const legendDatum of legendData) {
+                if (legendDatum.itemId === datum.itemId) {
+                    return { seriesId: legendDatum.seriesId, data: [{ ...legendDatum, enabled: visible }] };
+                }
+            }
+        }
+    }
+
+    public update(data?: CategoryLegendDatum[]) {
+        this.listeners.dispatch('legend-change', {
+            type: 'legend-change',
+            legendData: data ?? this.getData(),
+        });
+    }
+
+    public updateData(seriesId: string, data: CategoryLegendDatum[] = []) {
+        this.legendDataMap.set(seriesId, data);
+    }
+
+    public clearData() {
+        this.legendDataMap.clear();
+    }
+
+    public toggleItem({
+        enabled,
+        seriesId,
+        itemId,
+        legendItemName,
+    }: {
+        enabled: boolean;
+        seriesId: string;
+        itemId?: any;
+        legendItemName?: string;
+    }) {
+        if (legendItemName) {
+            this.getData().forEach((datum) => {
+                const newDatum = datum.legendItemName === legendItemName ? { ...datum, enabled } : datum;
+                this.updateData(datum.seriesId, [newDatum]);
+            });
+            return;
+        }
+
+        const seriesLegendData = this.getData(seriesId);
+        const singleLegendItem = seriesLegendData.length === 1;
+
+        const data = this.getData(seriesId).map((datum) =>
+            (itemId == null && singleLegendItem) || datum.itemId === itemId ? { ...datum, enabled } : datum
+        );
+
+        this.updateData(seriesId, data);
+    }
+
+    public getData(seriesId?: string) {
+        if (seriesId) {
+            return this.legendDataMap.get(seriesId) ?? [];
+        }
+
+        return [...this.legendDataMap].reduce(
+            (data, [_, legendData]) => data.concat(legendData),
+            [] as CategoryLegendDatum[]
+        );
```

Comment:
With the number of calls to `this.getData()`, it might be worth setting this value on `updateData()` and just accessing it directly internally instead of through the function?